### DISCUSSION TITLE
C# Support

### DISCUSSION THREAD

**Discussion by OeilDeLance** [bitbrain/beehave#221] [Q&A]
If I am not mistaken C# is not supported by beehave. I am currently investigating because we are going to make heavy use of BT's in Godot in the near future for a sizeable project.
My question: is C# support in discussion right now or is it not on the table at all ?
Another question would be: How hard/long would you estimate it would be to add it in? (for an experienced programmer for example).
Thank you !

**Comment by bitbrain** [comment]
In a project you should be able to mix and match .gd and C# without any issues: https://docs.godotengine.org/en/stable/tutorials/scripting/cross_language_scripting.html#:~:text=Godot%20allows%20you%20to%20mix,nodes%20written%20in%20different%20languages.

Is there anything in particular you are missing right now that does not work?

**Reply by OeilDeLance** [reply]
I must apologize I am completely new to Godot and didn't try much yet.
Are you saying that despite the lack of C# implementation examples of custom nodes (there is only gd script in the doc) it is entirely possible to write a custom beehave node in C# ?

Can I do that without having to write any GDScript ?
Or as your link suggests would we need to instantiate a c# script from CDScript for every custom Beehave node ?

**Reply by bitbrain** [reply]
As far as I am aware you can simply use any addon that is written in .gd script in C# and use it normally. No need to translate it.

**Reply by lostptr** [reply]
I mean, you can't extend, say, an ActionLeaf node and create your own action in C#. I was not sure about this so I just tried:
![image](https://github.com/bitbrain/beehave/assets/21160216/286ebcc5-cfd5-47e0-bb7b-a8804af84477)
The problem is that there is no way of inheriting from the ActionLeaf class because there isnt one in C#. 
So yeah, kinda sucks that you cant write a plugin that works with every language unless you write it in C++ I guess...

**Reply by bitbrain** [reply]
I was not aware of that! This is something that should be solved via Godot itself and not Beehave. Otherwise, any addon developer would need to re-create their entire addon in all various languages they wanted to support. The idea of Godot is that all languages have a common core API, so it should not matter to Godot what language a "type" is from, as types are something coming through Godot itself.

That being said, currently C# is baked somewhat into Godot as a module and is not using the GDExtension API, so any type exposed via C# won't be available as GDExtension types do.

I raised this issue in [the Godot contributor chat](https://chat.godotengine.org/channel/gdextension?msg=dcBNft9PHXHzMEySP).

**Reply by bitbrain** [reply]
**Update** this proposal would fix the issue: https://github.com/godotengine/godot-proposals/issues/7950
